Remove use of 2to3

rahulporuri opened the pull request:
Use `six` instead

~WARNING : I used `modernize` to make automated changes to the codebase. I then added the recommended changes without much review. As such, this PR might include changes which are not absolutely necessary to the functionality of the package. If necessary, I can review the changes and remove the unnecessary/overzealous changes.~

Commit:
FEAT : Remove use of 2to3

and use six instead

Commit:
FIX : Fixes

- Replace Long trait type with Int
- Do not modify dictionary during iteration

	modified:   apptools/io/h5/dict_node.py
	modified:   apptools/persistence/tests/test_state_pickler.py

rahulporuri commented:
The failures are now Python 2 only. See job https://travis-ci.org/enthought/apptools/jobs/531612677

Commit:
FIX : Pickle long types using StatePickler

Explicitly use long type in the type map used to pickle values in
StatePickler

	modified:   apptools/persistence/state_pickler.py
	new file:   apptools/persistence/tests/__init__.py
	modified:   apptools/persistence/tests/test_state_pickler.py

jwiggins reviewed (commented):
I think the blanket use of `from __future__ import absolute_import` is of dubious usefulness.

Additionally, it might be worth the time to examine each of the files where `six.moves` is being used. Specifically, `map`, `zip`, and `range` might be OK as is.

Commit:
CLN : Remove absolute_import imports from future

rahulporuri commented on apptools/template/impl/any_context_data_name_item.py:
```diff
@@ -122,13 +124,13 @@ def inputs_changed ( self ):
             filter = self.filter
```
looks like it. corrected.

rahulporuri commented on apptools/persistence/state_pickler.py:
```diff
@@ -440,7 +444,7 @@ def _do_list(self, value):
         idx = self._register(value)
         vals = [self._do(x) for x in value.values()]
-        data = dict(zip(value.keys(), vals))
+        data = dict(list(zip(list(value.keys()), vals)))
```
neither of them are. corrected.

rahulporuri commented on apptools/persistence/state_pickler.py:
```diff
@@ -270,7 +274,7 @@ def __init__(self):
         if PY_VER == 2:
             type_map[long] = self._do_basic_type
-            type_map[unicode] = self._do_basic_type
+            type_map[six.text_type] = self._do_basic_type
```
yes but this should not have been touched as it is python 2-only code.

```
>>> import six
>>> six.text_type
<type 'unicode'>
```

corranwebster commented on apptools/appscripting/script_manager.py:
```diff
@@ -23,11 +24,12 @@
+from .i_script_manager import IScriptManager
+from .lazy_namespace import add_to_namespace, FactoryWrapper, LazyNamespace
+from .scriptable_type import make_object_scriptable
+import six
```
Unfortunately modernize doesn't understand import order conventions :(

corranwebster commented on apptools/io/__init__.py:
```diff
@@ -14,4 +14,4 @@
     Part of the AppTools project of the Enthought Tool Suite.
 """
-from api import *
+from .api import *
```
We could at least make it explicit imports.

rahulporuri replied:
Made them full imports ("from apptools.io.api import *")

jwiggins replied:
I'm a little surprised that that doesn't cause a recursion error...

jwiggins replied:
(also, I was more horrified by the `*` than the relative import)

rahulporuri commented on apptools/lru_cache/tests/test_lru_cache.py:
```diff
@@ -141,7 +144,7 @@ def test_updated_event():
 
     c[0] = 0
-    print(c.keys())
+    print(list(c.keys()))
```
removing. maybe it was a left over debugging statement.

Commit:
CLN : Mostly move around imports

- and remove a debugging statement
- and remove an unnecessary use of six.text_type instead of unicode
- and remove an overzealous correction made by modernize
- and replace usage of sys python version with six

jwiggins commented on apptools/undo/tests/test_command_stack.py:
```diff
@@ -12,14 +12,15 @@
+from six.moves import range
 
+from traits.testing.unittest_tools import unittest
+from traits.testing.unittest_tools import unittest
```
Duped.

rahulporuri commented on apptools/io/h5/dict_node.py:
```diff
@@ -69,7 +69,7 @@ def __contains__(self, key):
 
     def keys(self):
-        return self.data.keys()
+        return list(self.data.keys())
```
This change needs discussion.

If we don't make this change, we have different behavior on Python 2 and Python 3.
If wrap the call with `list`, the return value of `H5DictNode.keys` differs from how a normal dictionary behaves.

jwiggins replied:
I think that because this method is part of the dictionary interface then it should act like a normal dictionary. Remove the `list` wrapper.

rahulporuri commented on apptools/lru_cache/lru_cache.py:
```diff
@@ -98,15 +98,15 @@ def get(self, key, default=None):
     def items(self):
         with self._lock:
-            return self._cache.items()
+            return list(self._cache.items())
```
Similar to an earlier question regd H5DictNode, should the items/keys/values methods return iterables on Python 3 or should they return lists?

jwiggins replied:
I want to say they should return iterables, but the locking presents a thorny problem. Perhaps the safest solution would be to implement this as an iterator (explicitly `yield`-ing items) on Python 3, and using the old/existing implementation on Python 2.

rahulporuri replied:
I'd prefer leaving this change as-is, mainly because we're interested in removing `LRUCache`. Is that okay?

Commit:
CLN : Remove unnecessary six.moves.map/range/zip

- and remove unnecessarily choosing between Int/Long trait types between
Python 2 and Python 3
- and fix an error where the import was renamed in a module but not all
usages were updated
- remove a few unnecessary six.iteritems calls and replace them with
dict.items. Except for in onecase where we were dealing with H5 files,
in which case, it stays iteritems

Commit:
CLN : Remove use of six.iteritems

and replace it with just calls to dict.items instead
and rename a few absolute imports to be full instead of implicit

	modified:   apptools/appscripting/script_manager.py
	modified:   apptools/io/__init__.py
	modified:   apptools/naming/__init__.py
	modified:   apptools/naming/adapter/__init__.py
	modified:   apptools/naming/trait_defs/__init__.py
	modified:   apptools/scripting/recorder.py
	modified:   apptools/type_manager/__init__.py

Commit:
DEV : Use warnings instead of printing warnings

	modified:   apptools/type_manager/adapter_manager.py

rahulporuri commented on apptools/persistence/tests/test_state_pickler.py:
```diff
@@ -5,17 +5,24 @@
+if six.PY2:
+    long_trait_type = Long
+else:
+    long_trait_type = Int
```
Hmm. I checked and the Int and Long don't share a common baseclass on Python 3. But they do use the same/similar validator for traits on Python 3.

Either way, this code isn't necessary. It was added when I was debugging an issue that this PR added! Removed these changes.

rahulporuri commented on apptools/io/h5/table_node.py:
```diff
@@ -1,6 +1,9 @@
 from pandas import DataFrame
 from tables.table import Table as PyTablesTable
+import six
+from six.moves import zip
```
The builtin `zip` is adequate. I removed this.

Commit:
CLN : Address comments in PR

- rewrite statement into a list comprehension instead of being wrapping
calls to map and zip with list multiple times

- return list/iterable on python 2/3 respectively, instead of returning
list on both 2/3.

	modified:   apptools/io/h5/dict_node.py
	modified:   apptools/io/h5/tests/test_file.py

rahulporuri commented:
@jwiggins @corranwebster this PR is ready for review. again. for the last time, i thnk.

rahulporuri commented on apptools/persistence/spickle.py:
```diff
@@ -15,6 +15,7 @@
+from __future__ import absolute_import
 import sys
 if sys.version_info[0] > 2:
     raise ImportError("This module does not work with Python 3.x")
```
I tried using the `exclude` kwarg when calling `find_packages` to exclude just this module from the Python 3 package but it looks like `exclude` will only help when entire packages/sub-packages need to be excluded and not when one specific module needs to be excluded.

jwiggins commented on apptools/io/h5/table_node.py:
```diff
@@ -1,6 +1,7 @@
 import numpy as np
 from pandas import DataFrame
 from tables.table import Table as PyTablesTable
+import six
```
Move this up so that it's after the pandas import

jwiggins commented on apptools/io/h5/tests/test_file.py:
```diff
@@ -509,15 +509,15 @@ def test_attribute_iteration_methods():
 
         # unfold the pairs
-        keys, vals = map(list, zip(*items))
+        keys, vals = [list(item) for item in zip(*items)]
```
No idea why this was originally written in such a complicated way. Just write `keys, vals = zip(*items)`. It's a common Python idiom and everything.

mdickinson replied:
FWIW, I'm really not a fan of the `zip(*stuff)` idiom: it's got a nasty not-immediately-obvious corner case where `stuff` is empty. Fine in this case, but then it gets re-used in places where it's not fine and people don't notice the corner case until it's too late.

```
>>> d = {'a': 1, 'b': 2}
>>> keys, values = zip(*d.items())  # fine
>>> d = {}
>>> keys, values = zip(*d.items())  # not fine
```

mdickinson replied:
This doesn't detract from @jwiggins' comment, of course.

One possible reason for the `list` is that otherwise `keys` and `values` are tuples. I haven't looked at the context, so I don't know if that matters in this particular case, but that's the usual reason for the extra `map(list, ...)` construct.

rahulporuri replied:
`keys`, `vals` end up as tuples when we just use `= zip(*items)`, which is not expected as we then try comparing to lists. I'm keeping the `[list(item) ... ]` code instead of updating other code to test against tuples or wrap `keys`/`vals` with `list` later in the code.

jwiggins replied:
Now that you've reminded me that `zip` returns tuples, that explains why it was initially written as `map(list, zip(*items))`.

jwiggins commented on apptools/lru_cache/tests/test_lru_cache.py:
```diff
@@ -73,7 +73,7 @@ def test_cache_resize():
 
 def test_cache_items():
     c = LRUCache(2)
-    assert_equal([], c.items())
```
`LRUCache.items` returns a `list` already.

jwiggins commented on apptools/lru_cache/tests/test_lru_cache.py:
```diff
@@ -105,15 +105,15 @@ def test_cache_idempotency():
 def test_cache_keys_values():
     c = LRUCache(2)
-    assert_equal([], c.items())
+    assert_equal([], list(c.items()))
```
`LRUCache.items` returns a `list` already.

jwiggins commented on apptools/naming/trait_defs/naming_traits.py:
```diff
@@ -27,6 +27,7 @@
 from apptools.naming.api \
     import Binding
+import six
 
```
This should probably be moved to above the ETS imports

jwiggins commented on apptools/naming/ui/context_node_type.py:
```diff
@@ -15,11 +15,13 @@
 
 # Enthought library imports.
+from __future__ import absolute_import
+from __future__ import print_function
```
Also the `# Enthought library imports.` comment above this is a bit off

jwiggins commented on apptools/preferences/preferences.py:
```diff
@@ -1,8 +1,8 @@
-
 # Standard library imports.
+from __future__ import absolute_import
+from __future__ import print_function
```
Useless diff now

jwiggins commented on apptools/preferences/tests/preferences_test_case.py:
```diff
@@ -164,14 +164,14 @@ def test_keys(self):
 
-        keys = p.keys()
+        keys = list(p.keys())
         keys.sort()
```
One liner: `keys = sorted(p.keys())`

jwiggins commented on apptools/preferences/ui/preferences_node.py:
```diff
@@ -15,11 +15,13 @@
 
 # Enthought library imports.
+from __future__ import absolute_import
+from __future__ import print_function
```
Not an Enthought library import

jwiggins commented on apptools/scripting/recorder.py:
```diff
@@ -221,9 +221,9 @@ def register(self, object, parent=None, trait_name_on_parent='',
+            sub_recordables = list(object.traits(record=True).keys())
             # Find all the trait names we must ignore.
-            ignore.extend(object.traits(record=False).keys())
+            ignore.extend(list(object.traits(record=False).keys()))
```
You can pass an iterable to `list.extend`

jwiggins commented on apptools/sweet_pickle/tests/updater_test_case.py:
```diff
@@ -238,8 +238,8 @@ def fn2():
         self.assertEqual(1, len(self.registry._state_function_classes))
-        self._validate_class_map_contents(updater.class_map.items() + \
-            updater2.class_map.items())
+        self._validate_class_map_contents(list(updater.class_map.items()) + \
```
The backtick is unnecessary

Commit:
CLN : Updated based on PR review

- remove unnecessary use of list
- move imports around to maintain consistency

	modified:   apptools/io/h5/table_node.py
	modified:   apptools/io/h5/tests/test_file.py
	modified:   apptools/lru_cache/tests/test_lru_cache.py
	modified:   apptools/naming/trait_defs/naming_traits.py
	modified:   apptools/naming/ui/context_node_type.py
	modified:   apptools/preferences/preferences.py
	modified:   apptools/preferences/tests/preferences_test_case.py
	modified:   apptools/preferences/ui/preferences_node.py
	modified:   apptools/scripting/recorder.py
	modified:   apptools/sweet_pickle/tests/updater_test_case.py

Commit:
CLN : More minor cleanup

- moving imports around and adding spaces between imports and
functions/classes
- remove unnecessary uses of list

	modified:   apptools/help/help_plugin/action/load_url_action.py
	modified:   apptools/help/help_plugin/help_code.py
	modified:   apptools/help/help_plugin/help_doc.py
	modified:   apptools/preferences/preferences.py
	modified:   apptools/preferences/tests/py_config_file.py
	modified:   apptools/template/impl/template_data_context.py
	modified:   apptools/type_registry/type_registry.py

Commit:
FIX : Revert earlier change

Re-add a list call which actually is needed as the output is a tuple
instead of a list

	modified:   apptools/io/h5/tests/test_file.py
